### PR TITLE
Rewrite headers between request and Template Repository response #29

### PR DESCRIPTION
**Pull request by wsatyla**
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: knotx-core/knotx-server/src/main/java/com/cognifide/knotx/server/KnotxServerVerticle.java [hunk 1]
```diff
@@ -90,10 +91,11 @@ public void start(Future<Void> fut) throws IOException, URISyntaxException {
                                                         }
                                                 );
                                     } else {
-                                        request.response().setStatusCode(HttpResponseStatus.NOT_FOUND.code()).end(repository.getReason());
+                                        rewriteHeaders(request, repository.getHeaders());
+                                        request.response().setStatusCode(repository.getStatusCode()).end();
                                     }
                                 },
-                                error -> LOGGER.error("Error: ", error)
+                                    error -> LOGGER.error("Error: ", error)
```
Comment (pawelkoper) [inline knotx-core/knotx-server/src/main/java/com/cognifide/knotx/server/KnotxServerVerticle.java]: please fix formatting

Comment (wsatyla) [inline knotx-core/knotx-server/src/main/java/com/cognifide/knotx/server/KnotxServerVerticle.java]: fixed

File: knotx-core/knotx-repository/src/main/java/com/cognifide/knotx/repository/impl/HttpRepository.java [hunk 2]
```diff
@@ -73,19 +70,20 @@ static HttpRepository of(String path, String domain, Integer port, Vertx vertx)
...
-    private Observable<RepositoryResponse> toRepositoryResponse(Buffer buffer) {
+    private Observable<RepositoryResponse> toRepositoryResponse(Buffer buffer, final HttpClientResponse httpClientResponse) {
         Observable<RepositoryResponse> response;
         if (buffer.length() > 0) {
             response = RepositoryResponse.success(buffer.toString()).toObservable();
         } else {
-            LOGGER.error("Remote repository returned empty template for path `{}`", path);
-            response = RepositoryResponse.error("No Template found for path %s", path).toObservable();
+            LOGGER.info("Remote repository returned empty template for path `{}` with status code {}", path, httpClientResponse.statusCode());
+            response = RepositoryResponse.error(httpClientResponse.statusCode(), httpClientResponse.headers()).toObservable();
```
Comment (marcinczeczko) [inline knotx-core/knotx-repository/src/main/java/com/cognifide/knotx/repository/impl/HttpRepository.java]: I think we can map this scenario to 404 response code. Repository returned empty HTML so for end user 404 is more readable

File: knotx-core/knotx-server/src/main/java/com/cognifide/knotx/server/KnotxServerVerticle.java [hunk 3]
```diff
@@ -79,6 +79,7 @@ public void start(Future<Void> fut) throws IOException, URISyntaxException {
                                                         result -> {
                                                             TemplateEngineResponse engineResponse = new TemplateEngineResponse(result.body());
                                                             if (engineResponse.isSuccess()) {
+                                                                rewriteHeaders(request, request.headers());
```
Comment (marcinczeczko) [inline knotx-core/knotx-server/src/main/java/com/cognifide/knotx/server/KnotxServerVerticle.java]: remove second parameter of rewriteHeaders() method. You already have request so headers can be taken inside the method.

File: knotx-core/knotx-server/src/main/java/com/cognifide/knotx/server/KnotxServerVerticle.java [hunk 4]
```diff
@@ -121,15 +123,22 @@ private JsonObject createEngineRequest(RepositoryResponse repositoryResponse, Ht
...
                 repositoryResponse.getData(),
                 request.method(),
-                getPreservedHeaders(request))
+                getPreservedHeaders(request.headers()))
                 .toJsonObject();
     }
 
-    private MultiMap getPreservedHeaders(HttpServerRequest request) {
+    private void rewriteHeaders(HttpServerRequest httpServerRequest, MultiMap headers) {
+        if (headers != null) {
```
Comment (marcinczeczko) [inline knotx-core/knotx-server/src/main/java/com/cognifide/knotx/server/KnotxServerVerticle.java]: And here you don't need to check for null headers - request will always return MultiMap even if no headers in actual requests

File: knotx-core/knotx-repository/src/main/java/com/cognifide/knotx/repository/impl/HttpRepository.java [hunk 5]
```diff
@@ -73,19 +70,20 @@ static HttpRepository of(String path, String domain, Integer port, Vertx vertx)
...
-                .reduce(Buffer::appendBuffer);
+                .reduce(Buffer::appendBuffer)
+                .flatMap(buffer -> toRepositoryResponse(buffer,response));
     }
 
-    private Observable<RepositoryResponse> toRepositoryResponse(Buffer buffer) {
+    private Observable<RepositoryResponse> toRepositoryResponse(Buffer buffer, final HttpClientResponse httpClientResponse) {
         Observable<RepositoryResponse> response;
         if (buffer.length() > 0) {
             response = RepositoryResponse.success(buffer.toString()).toObservable();
```
Comment (marcinczeczko) [inline knotx-core/knotx-repository/src/main/java/com/cognifide/knotx/repository/impl/HttpRepository.java]: What about success ? Headers should be also returned

File: knotx-core/knotx-server/src/main/java/com/cognifide/knotx/server/KnotxServerVerticle.java (hunk 4 above)
Comment (wsatyla) [inline knotx-core/knotx-server/src/main/java/com/cognifide/knotx/server/KnotxServerVerticle.java]: Fixed

Comment (wsatyla) [inline knotx-core/knotx-server/src/main/java/com/cognifide/knotx/server/KnotxServerVerticle.java]: Fixed

File: knotx-core/knotx-repository/src/main/java/com/cognifide/knotx/repository/impl/HttpRepository.java (hunk 5 above)
Comment (wsatyla) [inline knotx-core/knotx-repository/src/main/java/com/cognifide/knotx/repository/impl/HttpRepository.java]: Fixerd

Comment (marcinczeczko) [inline knotx-core/knotx-repository/src/main/java/com/cognifide/knotx/repository/impl/HttpRepository.java]: where it is fixed ? I don't see RepositoryResponse.success() consuming response headers here

**Comment by marcinczeczko** [comment]
Please also update integration tests to validate if headers are properly handled.

**Comment by Skejven** [comment]
Please add proper pull request message according to [Contributing Rules](https://github.com/Cognifide/knotx/blob/master/CONTRIBUTING.md#pull-requests).

File: knotx-core/knotx-repository/src/main/java/com/cognifide/knotx/repository/impl/HttpRepository.java (hunk 2 above)
Comment (wsatyla) [inline knotx-core/knotx-repository/src/main/java/com/cognifide/knotx/repository/impl/HttpRepository.java]: As we discussed. I've created an issue #48 to discuss how to react on other status codes than 200/400/500

**Comment by marcinczeczko** [comment]
Approved

**Pull request merged**